### DISCUSSION TITLE
Multiple threads possible if each uses own mg_mgr?

### DISCUSSION THREAD

**Discussion by numo68** [cesanta/mongoose#1527] [Ask questions, get answers] [answered]
The 6.x versions stated in `mongoose.h`

```
 * NOTE: Mongoose manager is single threaded. It does not protect
 * its data structures by mutexes, therefore all functions that are dealing
 * with a particular event manager should be called from the same thread,
 * with exception of the `mg_broadcast()` function. It is fine to have different
 * event managers handled by different threads.
```

The 7.x drops this and instead states the following in the documentation

```
NOTE: Since Mongoose's core is not protected against concurrent accesses, make
sure that all mg_* API functions are called from the same thread or RTOS task.
```

For some existing projects usages this is a _major_ difference and it is not always possible to have one manager (thread) to rule them all. So, was this change
1. an intentional one as there is some global data shared by multiple `mg_mgr`s, or
2. just a precaution warning that maybe some OS functions called might be not MT-safe at least on some platforms, or
3. the documentation is too strict and using one thread per `mg_mgr` is perfectly fine?

Thanks

**Comment by cpq** [comment, marked as answer]
Since recent changes, no global data is shared.
It is possible to have many threads with the `mg_mgr` in each. Answering your questions,
1. Up until recently, some code had global data - not anymore. All state is located in `mg_mgr`
2. IMO there are too many of them to mention
3. Yes, using one thread per mg_mgr is fine

The 6.x version used `mg_broadcast()` function for inter-thread comms. In 7.x, there is a [mg_mkpipe](https://mongoose.ws/documentation/#mg_mkpipe) / [mg_mgr_wakeup](https://mongoose.ws/documentation/#mg_mgr_wakeup)
 combo
